c2corg_api/security/discourse_client.py

```diff
@@ -1,71 +1,162 @@
 from pydiscourse.client import DiscourseClient
 
 import logging
-log = logging.getLogger(__name__)
 
+from base64 import b64encode, b64decode
+from pyramid.httpexceptions import HTTPBadRequest
+
+import hmac
+import hashlib
+import requests
+import urllib.error
+
+from urllib.parse import parse_qs
+
+log = logging.getLogger(__name__)
 
 # 10 seconds timeout for requests to discourse API
 # Using a large value to take into account a possible slow restart (caching)
 # of discourse.
-CLIENT_TIMEOUT = 10
+TIMEOUT = 10
+
+
+class APIDiscourseClient(object):
+
+    def __init__(self, settings):
+        self.settings = settings
+        self.discourse_base_url = settings['discourse.url']
+        self.discourse_public_url = settings['discourse.public_url']
+        self.api_key = settings['discourse.api_key']
+        self.sso_key = str(settings.get('discourse.sso_secret'))  # no unicode
+
+        self.discourse_userid_cache = {}
+        # FIXME: are we guaranteed usernames can never change? -> no!
+        self.discourse_username_cache = {}
+
+        self.client = DiscourseClient(
+                self.discourse_base_url,
+                api_username='system',  # the built-in Discourse user
+                api_key=self.api_key,
+                timeout=TIMEOUT)
+
+    def get_userid_by_userid(self, userid):
```

Comment:
Or simply: `get_userid` as the function will be called like `discouse_client.get_userid(someid)`.